### PR TITLE
New metric: Adjusted mutual info score

### PR DESCRIPTION
**Pull request by matsumotosan**
## What does this PR do?

Fixes part of #2003 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2058.org.readthedocs.build/en/2058/

### CODE REVIEW

**Comment by matsumotosan** [comment]
I have yet to implement `expected_mutual_info_score` as it is quite [involved](https://github.com/scikit-learn/scikit-learn/blob/7f9bad99d6e0a3e8ddf92a7e5561245224dab102/sklearn/metrics/cluster/_expected_mutual_info_fast.pyx).

File: src/torchmetrics/functional/clustering/adjusted_mutual_info_score.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    """Calculated expected mutual information score between two clusterings.
+
+    Args:
+        contingency: contingency matrix
+        n_samples: number of samples
+
+    Returns:
+        expected_mutual_info_score: expected mutual information score
+
+    """
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/adjusted_mutual_info_score.py:70]:
```suggestion
```
seems like it is not needed?

File: src/torchmetrics/clustering/adjusted_mutual_info_score.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        >>> ami_score(preds, target)
+        tensor(0.4744)
+
+    """
+
+    is_differentiable: bool = True
+    higher_is_better: Optional[bool] = None
+    full_state_update: bool = False
+    plot_lower_bound: float = 0.0
+    plot_upper_bound: float = 0.0
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/adjusted_mutual_info_score.py:73]:
```suggestion
    plot_upper_bound: float = 1.0
```

File: src/torchmetrics/clustering/adjusted_mutual_info_score.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+    - ``preds`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with predicted cluster labels
+    - ``target`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with ground truth cluster labels
+
+    As output of ``forward`` and ``compute`` the metric returns the following output:
+
+    - ``ami_score`` (:class:`~torch.Tensor`): A tensor with the Adjusted Mutual Information Score
+
+    Args:
+        average_method: Method used to calculate generalized mean for normalization
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/adjusted_mutual_info_score.py:55]:
```suggestion
        average_method: Method used to calculate generalized mean for normalization. Choose between
            ``'min'``, ``'geometric'``, ``'arithmetic'``, ``'max'``.
```

File: src/torchmetrics/clustering/adjusted_mutual_info_score.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    """
+
+    is_differentiable: bool = True
+    higher_is_better: Optional[bool] = None
+    full_state_update: bool = False
+    plot_lower_bound: float = 0.0
+    plot_upper_bound: float = 0.0
+    preds: List[Tensor]
+    target: List[Tensor]
+    contingency: Tensor
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/adjusted_mutual_info_score.py:76]:
```suggestion
```

File: src/torchmetrics/clustering/adjusted_mutual_info_score.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    Args:
+        average_method: Method used to calculate generalized mean for normalization
+        kwargs: Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics.clustering import AdjustedMutualInfoScore
+        >>> preds = torch.tensor([2, 1, 0, 1, 0])
+        >>> target = torch.tensor([0, 2, 1, 1, 0])
+        >>> ami_score = AdjustedMutualInfoScore("arithmetic")
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/adjusted_mutual_info_score.py:63]:
```suggestion
        >>> ami_score = AdjustedMutualInfoScore(average_method="arithmetic")
```

**Comment by matsumotosan** [comment]
@SkafteNicki @Borda CI is failing at "Install Sudo in container". Any idea what this is about?

**Comment by Borda** [comment]
> @SkafteNicki @Borda CI is failing at "Install Sudo in container". Any idea what this is about?

There is likely Microsoft outage, pls be patient and let's try it later again...

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
